### ISSUE TITLE
Provide progress or activity indication for repo clones

### ISSUE THREAD

**Issue by PythonNut** [raxod502/straight.el#97]
Cloning repositories can take a fair amount of time, so it would be good UI to demonstrate that the operation is still in-progress by either indicating progress or showing a spinner to reassure the user that their Emacs has not hung (as they occasionally do).

**Comment by raxod502** [comment]
I think the most reasonable way to do this would be to stream the output of `git clone` into the echo area. As far as I know, there isn't any other sort of "progress API" provided by Git.

**Comment by bbuccianti** [comment]
I'm starting to make some progress with this. Just wanted to ask something to you @raxod502 because I don't have a great knowledge of straight (and emacs).

Do you think that would be useful to change from call-process to something like async-start-process?

I mean, in order to display those "Counting objects: X%", etc, while it's happening we need to run git asynchronously, otherwise emacs it's frozen while the git process clone or do his stuff.

**Comment by raxod502** [comment]
Yes, `straight.el` needs to switch to asynchronous process invocation. However, it needs to be able to wait synchronously for a process to finish. This feature turns out to be extremely nontrivial, not because there's any theoretical reason it should be but rather because there are an alarming number of race conditions and hangs that can happen when you try to implement this feature. To my knowledge, there does not exist any other package that does it.

Please see #334, which I have attempted to solve on no fewer than three separate occasions. [Here](https://github.com/raxod502/straight.el/tree/wip/process-prompt) is the latest code, which already implements most of the work you would have to do. The only thing is to make sure that:

1. Under no circumstances does a series of process invocations cause an indefinite hang (some milliseconds to wait on a busy loop is OK)
2. Under no circumstances does subsequent code proceed without the process having finished properly first
3. The performance is not substantively worse than synchronous process invocation (losing some milliseconds is OK)

There are a number of different approaches depending on which Emacs standard library functions you use to wait on processes. It's not clear which of these actually should produce a difference in behavior, but clearly some of them do.

After #334 is implemented, it should be very easy to solve this issue as well with a minor change to the process filter function.
